app/models/service_retire_task.rb

```diff
@@ -32,31 +32,38 @@ def after_request_task_create
   def create_retire_subtasks(parent_service)
     parent_service.direct_service_children.each { |child| create_retire_subtasks(child) }
     parent_service.service_resources.collect do |svc_rsc|
-      next unless svc_rsc.resource.present? && svc_rsc.resource.respond_to?(:retire_now)
-      next if svc_rsc.resource.type.blank?
-      next if svc_rsc.resource_type == "ServiceTemplate" &&
-              !self.class.include_service_template?(self,
-                                                    svc_rsc.resource.id,
-                                                    parent_service)
+      next unless retireable?(svc_rsc, parent_service)
       nh = attributes.except("id", "created_on", "updated_on", "type", "state", "status", "message")
       nh['options'] = options.except(:child_tasks)
       # Initial Options[:dialog] to an empty hash so we do not pass down dialog values to child services tasks
       nh['options'][:dialog] = {}
-
-      new_task = (svc_rsc.resource.type.demodulize + "RetireTask").constantize.new(nh)
-      new_task.options.merge!(
-        :src_id              => svc_rsc.resource.id,
-        :service_resource_id => svc_rsc.id,
-        :parent_service_id   => parent_service.id,
-        :parent_task_id      => id,
-      )
-      new_task.source = svc_rsc.resource
-      new_task.request_type = svc_rsc.resource.type.demodulize.downcase + "_retire"
-      new_task.save!
+      new_task = create_task(svc_rsc, parent_service, nh)
       new_task.after_request_task_create
       miq_request.miq_request_tasks << new_task
       new_task.deliver_to_automate
       new_task
     end.compact!
   end
+
+  def retireable?(svc_rsc, parent_service)
+    srr = svc_rsc.resource
+    srr.present? &&
+      srr.respond_to?(:retire_now) &&
+      srr.type.present? &&
+      (svc_rsc.resource_type != "ServiceTemplate" || self.class.include_service_template?(self, srr.id, parent_service))
+  end
+
+  def create_task(svc_rsc, parent_service, nh)
+    new_task = (svc_rsc.resource.type.demodulize + "RetireTask").constantize.new(nh)
```

Comment:
Should this use `safe_constantize`?

Reply:
If there is an error the error message raised will be more meaningful if we use `constantize`.  With `safe_constantize` it will return a `nil` and and we will call `new` and the error message would be `undefined method 'new' for nil:NilClass`.

With `constantize` it will report the name of the invalid class.